xdsl/transforms/lower_csl_stencil.py

```diff
@@ -252,6 +255,123 @@ def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter,
                 block_arg.replace_by(SSAValue.get(new_arg))
 
 
+@dataclass(frozen=True)
+class FullStencilAccessImmediateReductionOptimization(RewritePattern):
+    """
+    If an apply op accesses all points in the stencil shape *and* immediately performs a reduction,
+    lower to an API call that iterates over all receive buffers at once. This requires setting up a
+    4d dsd that disregards all but one dimensions.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter, /):
+        # check that apply is inside a csl_wrapper and retreive `pattern` (stencil arm length + self)
+        if (wrapper := _get_module_wrapper(op)) is None:
+            return
+        pattern = wrapper.get_param_value("pattern").value.data
+
+        # get csl_stencil.access ops and offsets
+        access_ops: list[csl_stencil.AccessOp] = [
+            a for a in op.receive_chunk.walk() if isinstance(a, csl_stencil.AccessOp)
+        ]
+        offsets = set(tuple(a.offset) for a in access_ops)
+
+        # this rewrite only works if all points in the stencil shape are accessed
+        if not self.is_full_2d_starshaped_access(offsets, pattern - 1):
+            return
+
+        # find potential 'reduction' ops
+        reduction_ops: set[Operation] = set(
+            u.operation for a in access_ops for u in a.result.uses
+        )
+
+        # check if reduction ops are of the same type
+        red_op_ts = set(type(r) for r in reduction_ops)
+        if len(red_op_ts) > 1 or (red_op_t := red_op_ts.pop()) not in [
+            csl.FaddsOp,
+            csl.FmulsOp,
+        ]:
+            return
+        red_ops = cast(set[csl.BuiltinDsdOp], reduction_ops)
+
+        # safety check 1: each access has one use
+        acc_ops_uses = set(len(a.result.uses) for a in access_ops)
+        if len(acc_ops_uses) > 1 or acc_ops_uses.pop() != 1:
+            return
```

Comment:
```suggestion
        if any(len(a.result.uses) > 1 for a in access_ops):
            return
```

I think this is equivalent(?). I find it more readable (and it avoids constructing a `set`)

Reply:
Done as suggested.